clippy_lints/src/cfg_not_test.rs

```diff
@@ -0,0 +1,62 @@
+use clippy_utils::diagnostics::span_lint_and_help;
+use rustc_ast::NestedMetaItem;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Disallows usage of any conditional compilation that always excludes code from test builds.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Exclude code from tests builds. Is against simplicity in testing, guarding excessive mocking anti-pattern.
+    ///
+    /// Can show a codebase with a 100% coverage while having untested code.
+    ///
+    /// ### Example
+    /// ```rust
+    /// // an important that is a pain to test but not testing it can actually be dangerous
+    ///
+    /// #[cfg(not(test))]
+    /// important_check();
+    /// ```
+    ///
+    /// You should instead don't exclude any specific code from your test builds.
+    #[clippy::version = "1.73.0"]
+    pub CFG_NOT_TEST,
+    restriction,
+    "enforce against excluding code from test builds"
+}
+
+declare_lint_pass!(CfgNotTest => [CFG_NOT_TEST]);
+
+impl EarlyLintPass for CfgNotTest {
+    fn check_attribute(&mut self, cx: &EarlyContext<'_>, attr: &rustc_ast::Attribute) {
+        if let Some(ident) = attr.ident()
+            && ident.name == rustc_span::sym::cfg
+            && contains_not_test(attr.meta_item_list().as_deref(), false)
```

Comment:
This won't detect it if `--test` is passed.

This could be a pre-expansion pass lint instead so it's ran before `cfg` is expanded, *but these are very buggy and soft-deprecated*. See #6610 for an example of one of these issues. If it's made into a pre-expansion pass, this should probably be moved to `attrs.rs` and combined with `maybe_misused_cfg`, that way we can minimize its usage to only a few lints.

Whether or not it not detecting it if `--test` is passed is up to you, it's probably not as big of a deal as other lints.

Reply:
> I didn't know about `cfg` expansion, does this only concern `cfg_attr`?

No, it concerns `cfg` as well. In the HIR, and even AST, `cfg`s are expanded to either *nothing* or unchanged depending on whether it should "be in that compilation run". So, if `not(test)` is successful, it'll be entirely gone before linting occurs. Thus, if the user passes `--test`, there will be no warnings.

For something like `not(test)`, not a big deal, but something like `cfg(all(windows, not(test))` is not good.

I'd recommend messing around a bit on godbolt if you're still having trouble understanding: https://godbolt.org/z/WxPWeM416, you can see the AST before macro-expansion, after, and the HIR by passing different things to `unpretty`.

> (btw, why isn't there more folders for lints?)

Only because nobody's taken the time to :D I'd be 100% in favor of merging a PR to clean that up a bit.

> I don't know if grouping lints has an advantage

It's cleaner, but more importantly, it puts them in the same lint pass which has a *huge* performance boost. For on-by-default (warn/deny) lints, that you almost never want to globally disallow, this is great. In the case of this lint, it's allow-by-default, but I'd still say it's ok in the same lint pass as it does the exact same thing as other lints.

> Also, this deserves his own lint, and IMHO has not much to do with `maybe_misused_cfg`.

I actually agree now; that one is warn-by-default and this one is a restriction lint. So, no complaints from me.

Reply:
> merging a PR to clean that up a bit.
Would definitely be willing to do that 😉 

> In the case of this lint, it's allow-by-default
I thought restriction was warn-by-default (it's not). In the original issue, @blyxyas states "I feel like the drawback is too heavy for an allow-by-default lint" which I understand as "the lint should be warn-by-default"

Reply:
I think she meant enabled-by-default, `restriction` is allow-by-default and generally, lints in that group aren't applicable in every situation or go against idiomatic code. There are valid reasons for using `not(test)`.